File: src/main/scala/com/sageserpent/americium/java/TrialsTestExtension.java
package com.sageserpent.americium.java;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;
import com.google.common.collect.Lists;
import cyclops.companion.Streams;
import cyclops.data.tuple.Tuple2;
import cyclops.data.tuple.Tuple3;
import cyclops.data.tuple.Tuple4;
import lombok.Value;
import org.junit.jupiter.api.extension.*;
import org.junit.platform.commons.support.AnnotationSupport;
import org.junit.platform.commons.support.HierarchyTraversalMode;
import org.junit.platform.commons.support.ReflectionSupport;
import org.junit.platform.commons.util.ExceptionUtils;
import org.opentest4j.TestAbortedException;

import java.lang.invoke.MethodType;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class TrialsTestExtension
        implements TestTemplateInvocationContextProvider {
    private final static Class<? extends Throwable>[]
            additionalExceptionsToHandleAsFiltration =
            ImmutableSet
                    .of(TestAbortedException.class)
                    .stream()
                    .toArray(Class[]::new);
    private final static ImmutableList<TupleAdaptation<?>>
            tupleAdaptations =
            ImmutableList.of(new TupleAdaptation<>(Tuple2.class,
                                                   (tuple) -> Collections.unmodifiableList(
                                                           Arrays.asList(
                                                                   tuple.toArray()))),
                             new TupleAdaptation<>(Tuple3.class,
                                                   (tuple) -> Collections.unmodifiableList(
                                                           Arrays.asList(
                                                                   tuple.toArray()))),
                             new TupleAdaptation<>(Tuple4.class,
                                                   (tuple) -> Collections.unmodifiableList(
                                                           Arrays.asList(
                                                                   tuple.toArray()))));


    private static Iterator<TestIntegrationContext<Object>> testIntegrationContexts(
            ExtensionContext context) {
        final Method testMethod = context.getRequiredTestMethod();

        return AnnotationSupport
                .findAnnotation(testMethod, TrialsTest.class)
                .map(annotation -> {
                    final List<Trials<Object>> trials =
                            Lists.transform(instancesReferredToBy(Stream
                                                                          .of(annotation.trials())
                                                                          .collect(
                                                                                  Collectors.toList()),
                                                                  context,
                                                                  TrialsScaffolding.class),
                                            TrialsScaffolding::trials);

                    final Trials<List<Object>> lists = Trials
                            .api()
                            .collections(trials,
                                         () -> new Builder<Object,
                                                 List<Object>>() {
                                             // This builder tolerates null
                                             // elements, which is why we use
                                             // `.collections` here instead
                                             // of `.lists`, which would seem
                                             // to be the natural  choice.
                                             final List<Object> buffer =
                                                     new LinkedList<>();

                                             @Override
                                             public void add(Object caze) {
                                                 buffer.add(caze);
                                             }

                                             @Override
                                             public List<Object> build() {
                                                 return Collections.unmodifiableList(
                                                         buffer);
                                             }
                                         });

                    return (TrialsScaffolding.SupplyToSyntax) lists
                            .withLimits(annotation.casesLimit(),
                                        TrialsScaffolding.OptionalLimits
                                                .builder()
                                                .complexity(annotation.complexity())
                                                .shrinkageAttempts(
                                                        annotation.shrinkageAttempts())
                                                .build());
                })
                .orElseGet(() -> AnnotationSupport
                        .findAnnotation(testMethod,
                                        ConfiguredTrialsTest.class)
                        .map(annotation -> instancesReferredToBy(
                                Stream
                                        .of(annotation.value())
                                        .collect(
                                                Collectors.toList()),
                                context,
                                TrialsScaffolding.SupplyToSyntax.class).get(
                                0))
                        .orElseThrow(() -> {
                            throw new TestAbortedException(String.format(
                                    "`TrialsTest` annotation missing from" +
                                    " " +
                                    "method: %s",
                                    testMethod));
                        })).testIntegrationContexts();
    }

    private static <Clazz> List<Clazz> instancesReferredToBy(
            List<String> supplierFieldNames, ExtensionContext context,
            Class<? extends Clazz> clazz) {

        final Class<?> testClass = context.getRequiredTestClass();

        final List<Field> supplierFields =
                ReflectionSupport.findFields(testClass,
                                             field -> supplierFieldNames.contains(
                                                     field.getName()),
                                             HierarchyTraversalMode.BOTTOM_UP);

        final Map<String, Field> fieldsByName = supplierFields
                .stream()
                .filter(field -> clazz.isAssignableFrom(field.getType()))
                .collect(Collectors.toMap(Field::getName, Function.identity()));


        final Object testInstance = context.getTestInstance().orElse(null);

        return supplierFieldNames.stream().map(fieldName -> {
            try {
                final Field candidateField =
                        fieldsByName.get(fieldName);
                return (Clazz) Optional
                        .ofNullable(candidateField)
                        .flatMap(field -> ReflectionSupport
                                .tryToReadFieldValue(field,
                                                     testInstance)
                                .toOptional())
                        .orElseThrow(() ->
                                             supplierFields
                                                     .stream()
                                                     .filter(field -> fieldName.equals(
                                                             field.getName()))
                                                     .findFirst()
                                                     .map(field -> new RuntimeException(
                                                             String.format(
                                                                     "Field " +
                                                                     "of name" +
                                                                     " `%s` " +
                                                                     "in test" +
                                                                     " class " +
                                                                     "`%s` " +
                                                                     "has the" +
                                                                     " wrong " +
                                                                     "type of" +
                                                                     " `%s` -" +
                                                                     " should" +
                                                                     " be " +
                                                                     "typed " +
                                                                     "as a %s.",
                                                                     fieldName,
                                                                     testClass,
                                                                     field.getType(),
                                                                     clazz)))
                                                     .orElse(new RuntimeException(
                                                             String.format(
                                                                     "Failed " +
                                                                     "to find" +
                                                                     " field " +
                                                                     "of name: `%s` in test class `%s`.",
                                                                     fieldName,
                                                                     testClass)))
                        );
            } catch (Exception e) {
                ExceptionUtils.throwAsUncheckedException(e);
                return null;
            }
        }).collect(Collectors.toList());
    }

    private static List<Object> wrap(Object listOrSingleItem) {
        if (listOrSingleItem instanceof List)
            return (List<Object>) listOrSingleItem;
        else {
            final List<Object>
                    buffer =
                    new LinkedList<>();
            buffer.add(listOrSingleItem);

            return Collections.unmodifiableList(
                    buffer);
        }
    }

    @Override
    public boolean supportsTestTemplate(ExtensionContext context) {
        return true;
    }

    @Override
    public Stream<TestTemplateInvocationContext> provideTestTemplateInvocationContexts(
            ExtensionContext context) {
        final Method method = context.getRequiredTestMethod();
        final Class<?>[] formalParameterTypes = method.getParameterTypes();

        return Streams
                .stream(testIntegrationContexts(context))
                .map(testIntegrationContext -> new TestTemplateInvocationContext() {
                    @Override
                    public String getDisplayName(int invocationIndex) {
                        final String shrinkagePrefix =
                                testIntegrationContext.isPartOfShrinkage()
                                ? "Shrinking ... " : "";

                        final List<Object> caze =
                                wrap(testIntegrationContext.caze());
                        return String.format("%s%s %s",
                                             shrinkagePrefix,
                                             TestTemplateInvocationContext.super.getDisplayName(
                                                     invocationIndex),
                                             1 <
                                             caze.size()
                                             ? caze
                                             :
                                             caze.get(0));
                    }

                    @Override
                    public List<Extension> getAdditionalExtensions() {
                        final List<Object> adaptedArguments =
                                new LinkedList<>();

                        final Map<Integer, TupleAdaptation<Object>>
                                cachedTupleAdaptations = new HashMap<>();

                        {
                            // Behold, Java in all its glory....
                            final AtomicInteger formalParameterIndex =
                                    new AtomicInteger(0);

                            final Iterator<Object> argumentIterator =
                                    wrap(testIntegrationContext.caze()).iterator();

                            while (formalParameterTypes.length >
                                   formalParameterIndex.get() &&
                                   argumentIterator.hasNext()) {
                                final Object argument =
                                        argumentIterator.next();
                                final Class<?> formalParameterType =
                                        formalParameterTypes[formalParameterIndex.get()];

                                final List<Object> expansion =
                                        cachedTupleAdaptations.computeIfAbsent(
                                                formalParameterIndex.get(),
                                                unused ->
                                                        formalParameterType.isInstance(
                                                                argument) ?
                                                        (TupleAdaptation<Object>) TupleAdaptation.fallbackForNonTuples
                                                                          :
                                                        (TupleAdaptation<Object>) tupleAdaptations
                                                                .stream()
                                                                .filter(tupleAdaptation -> tupleAdaptation.type.isInstance(
                                                                        argument))
                                                                .findFirst()
                                                                .orElse(TupleAdaptation.fallbackForNonTuples)).expansion.apply(
                                                argument);
                                formalParameterIndex.addAndGet(expansion.size());
                                adaptedArguments.addAll(expansion);
                            }
                        }

                        return ImmutableList.of(new ParameterResolver() {
                            @Override
                            public boolean supportsParameter(
                                    ParameterContext parameterContext,
                                    ExtensionContext extensionContext)
                                    throws ParameterResolutionException {
                                return Optional
                                        .ofNullable(adaptedArguments
                                                            .get(parameterContext.getIndex()))
                                        .map(parameter -> {
                                            final Class<?> formalParameterType =
                                                    parameterContext
                                                            .getParameter()
                                                            .getType();
                                            final Class<?>
                                                    formalParameterReferenceType =
                                                    formalParameterType.isPrimitive()
                                                    ? MethodType
                                                            .methodType(
                                                                    formalParameterType)
                                                            .wrap()
                                                            .returnType()
                                                    : formalParameterType;
                                            return formalParameterReferenceType.isInstance(
                                                    parameter);
                                        })
                                        .orElse(true);
                            }

                            @Override
                            public Object resolveParameter(
                                    ParameterContext parameterContext,
                                    ExtensionContext extensionContext)
                                    throws ParameterResolutionException {
                                return adaptedArguments.get(
                                        parameterContext.getIndex());
                            }
                        }, new InvocationInterceptor() {
                            @Override
                            public void interceptTestTemplateMethod(
                                    Invocation<Void> invocation,
                                    ReflectiveInvocationContext<Method> invocationContext,
                                    ExtensionContext extensionContext) {
                                if (!testIntegrationContext
                                        .inlinedCaseFiltration()
                                        .executeInFiltrationContext(
                                                () -> {
                                                    try {
                                                        InvocationInterceptor.super.interceptTestMethod(
                                                                invocation,
                                                                invocationContext,
                                                                extensionContext);
                                                    } catch (Throwable throwable) {
                                                        ExceptionUtils.throwAsUncheckedException(
                                                                throwable);
                                                    }
                                                },
                                                additionalExceptionsToHandleAsFiltration))
                                    throw new TestAbortedException();
                            }
                        }, new TestWatcher() {
                            @Override
                            public void testFailed(ExtensionContext context,
                                                   Throwable cause) {
                                testIntegrationContext
                                        .caseFailureReporting()
                                        .report(cause);
                            }
                        });
                    }
                });
    }

    @Value
    static class TupleAdaptation<Tuple> {
        public static TupleAdaptation<?> fallbackForNonTuples =
                new TupleAdaptation<>(Object.class, Collections::singletonList);

        Class<Tuple> type;
        Function<Tuple, List<Object>> expansion;
    }
}

File: src/main/scala/com/sageserpent/americium/java/TrialsTestExtension.scala
package com.sageserpent.americium.java
import com.google.common.collect.ImmutableList
import com.sageserpent.americium.java.SoonToBeRemovedTrialsTestExtension.testIntegrationContexts
import cyclops.companion.Streams
import cyclops.data.tuple.{
  Tuple2 as JavaTuple2,
  Tuple3 as JavaTuple3,
  Tuple4 as JavaTuple4
}
import org.junit.jupiter.api.extension.*
import org.junit.platform.commons.util.ExceptionUtils
import org.opentest4j.TestAbortedException

import java.lang.invoke.MethodType
import java.lang.reflect.Method
import java.util
import java.util.Collections
import java.util.stream.Stream

class TrialsTestExtension extends SoonToBeRemovedTrialsTestExtension {
  private val additionalExceptionsToHandleAsFiltration
      : Array[Class[_ <: Throwable]] =
    Set(classOf[TestAbortedException]).toArray

  private def wrap(listOrSingleItem: Any) = if (
    listOrSingleItem.isInstanceOf[util.List[_]]
  ) listOrSingleItem.asInstanceOf[util.List[Any]]
  else Collections.singletonList(listOrSingleItem)

  case class TupleAdaptation[Tuple](
      clazz: Class[Tuple],
      expansion: Tuple => util.List[Any]
  )

  val fallbackForNonTuples: TupleAdaptation[Any] =
    TupleAdaptation(classOf[Any], Collections.singletonList[Any])

  protected val tupleAdaptations: ImmutableList[TupleAdaptation[_]] =
    ImmutableList.of(
      new TupleAdaptation[JavaTuple2[_, _]](
        classOf[JavaTuple2[_, _]],
        (tuple: JavaTuple2[_, _]) =>
          Collections.unmodifiableList(util.Arrays.asList(tuple.toArray: _*))
      ),
      new TupleAdaptation[JavaTuple3[_, _, _]](
        classOf[JavaTuple3[_, _, _]],
        (tuple: JavaTuple3[_, _, _]) =>
          Collections.unmodifiableList(util.Arrays.asList(tuple.toArray: _*))
      ),
      new TupleAdaptation[JavaTuple4[_, _, _, _]](
        classOf[JavaTuple4[_, _, _, _]],
        (tuple: JavaTuple4[_, _, _, _]) =>
          Collections.unmodifiableList(util.Arrays.asList(tuple.toArray: _*))
      )
    )

  override def supportsTestTemplate(context: ExtensionContext) = true

  override def provideTestTemplateInvocationContexts(
      context: ExtensionContext
  ): Stream[TestTemplateInvocationContext] = {
    val method               = context.getRequiredTestMethod
    val formalParameterTypes = method.getParameterTypes
    Streams
      .stream(testIntegrationContexts(context))
      .map((testIntegrationContext: TestIntegrationContext[Any]) =>
        new TestTemplateInvocationContext() {
          override def getDisplayName(invocationIndex: Int): String = {
            val shrinkagePrefix =
              if (testIntegrationContext.isPartOfShrinkage) "Shrinking ... "
              else ""
            val caze = wrap(testIntegrationContext.caze)
            String.format(
              "%s%s %s",
              shrinkagePrefix,
              super.getDisplayName(invocationIndex),
              if (1 < caze.size) caze
              else caze.get(0)
            )
          }

          override def getAdditionalExtensions: util.List[Extension] = {
            val adaptedArguments = new util.LinkedList[Any]
            val cachedTupleAdaptations =
              new util.HashMap[Integer, TupleAdaptation[
                Any
              ]]
            var formalParameterIndex = 0
            val argumentIterator = wrap(testIntegrationContext.caze).iterator
            while ({
              formalParameterTypes.length > formalParameterIndex && argumentIterator.hasNext
            }) {
              val argument = argumentIterator.next
              val formalParameterType =
                formalParameterTypes(formalParameterIndex)
              val expansion = cachedTupleAdaptations
                .computeIfAbsent(
                  formalParameterIndex,
                  (unused: Integer) =>
                    if (formalParameterType.isInstance(argument))
                      fallbackForNonTuples
                        .asInstanceOf[TupleAdaptation[
                          Any
                        ]]
                    else
                      tupleAdaptations.stream
                        .filter(
                          (tupleAdaptation: TupleAdaptation[
                            _
                          ]) => tupleAdaptation.clazz.isInstance(argument)
                        )
                        .findFirst
                        .orElse(fallbackForNonTuples)
                        .asInstanceOf[TupleAdaptation[
                          Any
                        ]]
                )
                .expansion
                .apply(argument)
              formalParameterIndex += expansion.size
              adaptedArguments.addAll(expansion)
            }

            ImmutableList.of(
              new ParameterResolver() {
                override def supportsParameter(
                    parameterContext: ParameterContext,
                    extensionContext: ExtensionContext
                ): Boolean = Option(
                  adaptedArguments.get(parameterContext.getIndex)
                ).forall((parameter: Any) => {
                  val formalParameterType =
                    parameterContext.getParameter.getType
                  val formalParameterReferenceType =
                    if (formalParameterType.isPrimitive)
                      MethodType
                        .methodType(formalParameterType)
                        .wrap
                        .returnType
                    else formalParameterType
                  formalParameterReferenceType.isInstance(parameter)
                })
                override def resolveParameter(
                    parameterContext: ParameterContext,
                    extensionContext: ExtensionContext
                ): Any = adaptedArguments.get(parameterContext.getIndex)
              },
              new InvocationInterceptor() {
                override def interceptTestTemplateMethod(
                    invocation: InvocationInterceptor.Invocation[Void],
                    invocationContext: ReflectiveInvocationContext[Method],
                    extensionContext: ExtensionContext
                ): Unit = {
                  if (
                    !testIntegrationContext.inlinedCaseFiltration
                      .executeInFiltrationContext(
                        () => {
                          try
                            super.interceptTestMethod(
                              invocation,
                              invocationContext,
                              extensionContext
                            )
                          catch {
                            case throwable: Throwable =>
                              ExceptionUtils
                                .throwAsUncheckedException(throwable)
                          }
                        },
                        additionalExceptionsToHandleAsFiltration
                      )
                  ) throw new TestAbortedException
                }
              },
              new TestWatcher() {
                override def testFailed(
                    context: ExtensionContext,
                    cause: Throwable
                ): Unit = {
                  testIntegrationContext.caseFailureReporting.report(cause)
                }
              }
            )
          }
        }
      )
  }
}
